### PR TITLE
added class KeepBest to address Issue #2768

### PR DESCRIPTION
**Pull request by alvarouc**
### Summary

New callback class to track best model according to best validation metric. It stores the model weights on memory instead of saving the model at each epoch. This saves significant time and removes disk utilization for returning the best model.

### Related Issues

#2768

### PR Overview

- [y] This PR requires new unit tests [y/n] (make sure tests are included)
- [y] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [y] This PR is backwards compatible [y/n]
- [n] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

I apologize for the unwanted changes besides the addition of the new class. Instead of fighting my IDE I hoped we could cherry-pick this one.

### CODE REVIEW

**Comment by lucasdavid** [comment]
The behavior you're describing is already implemented in the `EarlyStopping` callback, when`restore_best_weights=True` is passed.
If you do not wish to early stop, I supposed you could initiate it with `patience=epochs` as well.

**Comment by louis925** [comment]
> The behavior you're describing is already implemented in the `EarlyStopping` callback, when`restore_best_weights=True` is passed.
> If you do not wish to early stop, I supposed you could initiate it with `patience=epochs` as well.

Does the EarlyStopping callback write the weights to disk or memory?

**Comment by louis925** [comment]
> The behavior you're describing is already implemented in the `EarlyStopping` callback, when`restore_best_weights=True` is passed.
> If you do not wish to early stop, I supposed you could initiate it with `patience=epochs` as well.

I think the current implementation of the `EarlyStop` with `restore_best_weights=True` is different from what we have here. The `EarlyStop` will not restore the best weight when training end if the wait epochs does not exceed `patience`. But here, `KeepBest` class always restores the best weight at the end of training.

**Comment by lucasdavid** [comment]
That's true! I haven't thought of that.

**Comment by fchollet** [comment]
Thank you for preparing this PR.

We are no longer adding new features to multi-backend Keras, as we are refocusing development efforts on `tf.keras`. If you are still interested in submitting this PR, please direct it to `tf.keras` in the TensorFlow repository instead.
